### PR TITLE
Add login entrypoint

### PR DESCRIPTION
**Pull request by sgomes**
This change extracts login into its own entrypoint, with the minimal required set of reducers and dependencies. This greatly reduces the amount of JS (and thus, time) required to boot `/log-in`.

This PR is largely based off of @jsnajdr's #32927.

This PR *appears* to work, but I'm not sure how to test the functionality extensively, beyond attempting login with 2FA and generating/using a magic link. Those all work, and so do E2E tests. I'm not sure what else to look for, which is where hopefully you come in! Thank you! 🙇 

#### Changes proposed in this Pull Request

* Create new entrypoint for login.
* Add login entrypoint bootup files, including templates and store initialisation.
* Remove login from regular sections list.
* Extract common utilities from `client/controller/index.web.js` to `client/controller/util-web.js`.
* Extract `ui.section` and `ui.masterbarVisibility` reducers into their own modules.

#### Testing instructions

* Test as much login functionality as possible with the live branch. Magic login, wpcom login, 2FA, etc., should all be tested.

### CODE REVIEW

**Comment by sgomes** [comment]
The numbers seem to be pretty different from what I was getting. Looking into that.

**Comment by sgomes** [comment]
In its current state, this PR cuts the JS size of `/log-in` from ~700KB to ~370KB, and the total size of `/log-in` from ~830KB to ~470KB.

**Comment by sgomes** [comment]
Rebasing on master to pick up latest changes to E2E tests. This will hopefully stop them failing.

**Comment by sgomes** [comment]
It appears that the E2E tests are failing on magic login. Investigating.

**Comment by sgomes** [comment]
The magic login issues appear to have been fixed, but E2E tests are still failing.

**Comment by sgomes** [comment]
Looks like it was flakiness. E2E tests pass now 🎉

**Comment by sgomes** [comment]
Removed the draft status, as I believe this is now ready for review.

@blowery Any idea what team I should add for making sure login doesn't break?

**Comment by sgomes** [comment]
Rebased on master to pick up `lodash-es` change. This may further reduce the size of the new chunks.

**Comment by sgomes** [comment]
The Woo E2E canary is consistently failing. Investigating.

**Comment by sgomes** [comment]
The issues may be related to the fact that SSR isn't working correctly for login. Investigating.

**Comment by sgomes** [comment]
Rebased on master to pull in changes from #33804.

**Comment by sgomes** [comment]
SSR should now be working correctly. Previously broken E2E tests are now passing. They still failed, but it appears to be due to flakiness. Re-running to make sure.

**Comment by sgomes** [comment]
Failures in E2E tests appear to be flakiness, as it fails in different places during different runs. 

@bsessions85: Any idea if this is expected, or if this PR is somehow introducing additional flakiness?

**Comment by bsessions85** [comment]
@sgomes  You're good one this one. The e2e tests were just flaky for some reason. I'll keep watching on other builds but I would say they are fine for this.

**Comment by sgomes** [comment]
Thanks for looking into that, @bsessions85 ! 🙂

File: client/landing/login/store.js
Comment (jsnajdr) [inline client/landing/login/store.js]: You also need to add the `users` reducer for `currentUser` to work. There is only `state.currentUser.id`, the rest of the user object is in `state.users.items[ state.currentUser.id ]`. I'd like to change that in the future, I even have a WIP branch for that, but it's not trivial.

But the user is not initialized anyway -- the login page always shows the logged-out version. Because the [user initialization code](https://github.com/Automattic/wp-calypso/blob/master/client/boot/app.js#L38-L45) is missing in the login entrypoint.

I think that showing just one version of the login page, no matter whether user is already logged in or not, is actually a good thing to do. It would simplify the login page a lot. The only observable difference between them right now is the masterbar.

But that should probably be a separate and intentional patch.

File: client/landing/login/index.js
Comment (jsnajdr) [inline client/landing/login/index.js]: The `context.store` is already done in the `setupMiddlewares` function, both on enter and exit. Creating small focused page.js handlers is a good direction, so I'd recomment removing the code from `setupMiddlewares`.

File: client/landing/login/index.js
Comment (jsnajdr) [inline client/landing/login/index.js]: Instead of this special handler, just pass an enhanced router to `initLoginSection`:
```js
initLoginSection( ( route, ...handlers ) => page( route, ...handlers, renderHandler ) );
function renderHandler( context, next ) {
  ( context.serverSideRender ? hydrate : render )( context );
  next();
}
```

Context: the section init function is passed a `router` object as argument, and inside it, you can setup route handlers either with `router` or `page`:
```js
export default router => {
  page( '/x', showX, makeLayout, clientRender );
  router( '/y', showY, makeLayout );
}
```
Note that `clientRender` is missing in the `router` call. That's because it's added automatically at the end, and it's Express.js or page.js function, depending on where the code runs.

The `router` thing is intended for isomorphic pages. See also the [`clientRouter`](https://github.com/Automattic/wp-calypso/blob/7aaf16531bb4d948e6b49b6df14e19e580354650/client/controller/index.web.js#L61-L63) function and its [usage in the section loader](https://github.com/Automattic/wp-calypso/blob/7aaf16531bb4d948e6b49b6df14e19e580354650/client/sections-middleware.js#L40).

File: client/components/global-notices/index.jsx
Comment (sgomes) [reply client/components/global-notices/index.jsx]: Done! :)

File: client/landing/login/store.js
Comment (sgomes) [reply client/landing/login/store.js]: Interesting, that is indeed confusing.

It did remind me of something, though: the user should get redirected if they're already logged in, so there shouldn't be any need for this reducer in the first place inside of the login section. I'm attempting to remove it and running through the E2E test suite to see if that works.

Comment (jsnajdr) [reply client/landing/login/store.js]: > the user should get redirected if they're already logged in

That currently doesn't happen though. And for the case when I want to log in as someone else, a more friendly UI would be to display a "Already logged in as X.Y. Is it you?" warning.

File: client/landing/login/index.js
Comment (sgomes) [reply client/landing/login/index.js]: Done!

File: client/landing/login/store.js
Comment (sgomes) [reply client/landing/login/store.js]: The first attempt failed, since the `getCurrentUserId` selector was returning `undefined`, and there are places in the code where `null` is explicitly expected for logged out users. Modified the `getCurrentUserId` selector to default to `null`. Waiting for E2E results to come back.

File: client/landing/login/index.js
Comment (sgomes) [reply client/landing/login/index.js]: Done! Thanks for the suggestion!

File: client/landing/login/store.js
Comment (sgomes) [reply client/landing/login/store.js]: > That currently doesn't happen though. And for the case when I want to log in as someone else, a more friendly UI would be to display a "Already logged in as X.Y. Is it you?" warning.

In the midst of all the changes I got confused. You're right, that's not the current behaviour.

So as I see it, we have some options here:
1 - You get the `log-in` page without any information on whether you're already logged in.
2 - (current) You get the `log-in` page with the logged in masterbar on top, that shows a not so obvious indication you're already logged-in
3 - You get the `log-in` page with the logged in masterbar on top as well as a clearer indication you're already logged in, e.g. "Already logged in as X.Y. Is it you?"

As I understand it, you're arguing for option 2 for now, with option 3 in a later PR?

In the meantime, I'll run some tests to see what the difference in bundle size would be between having user initialization code and not.

Comment (sgomes) [reply client/landing/login/store.js]: So, supporting the logged in layout (`client/layout/index.jsx`) would pretty much negate all the savings in this PR, as it pulls in the world and requires a bunch more reducers to be set up.

The alternative would be to create a customized layout which is essentially the logged out layout + the logged in masterbar. That would likely result in an increase of 1-2KB vs. always logged out because of the extra user initialization and user reducers.

How do you want me to move forward, @jsnajdr ?

Comment (jsnajdr) [reply client/landing/login/store.js]: I see two good options:
1. showing exactly the same login page for both logged in and logged out users. The masterbar doesn't indicate logged-in status. A downside is that already logged-in user is in a dead end here: there are no links to a logged-in view from the login page. The logged-in masterbar provided these links.
2. redirecting automatically to a logged-in home page. That's what Dropbox does, for example. The downside is that I can't easily login to another account. I need to go through the `/me` section and click `Sign Out` there. Few more clicks and you need to know where the "Sign Out" is.

Good amendment to option 1 would be detecting that I'm already logged in and offering to proceed or to use another account. That's what Google does. We don't need to bundle the huge `layout/logged-in` to achieve that: just issuing the `/me` request and storing the result in `state.currentUser` should be sufficient. 1-2KB is a good code size estimate.

I don't have enough domain and historical knowledge about our login to make recommendations. Let's ping @stephanethomas and @scruffian to see if they have any insights that could help.

Comment (blowery) [reply client/landing/login/store.js]: I think @westi had opinions on these flows too.

Comment (stephanethomas) [reply client/landing/login/store.js]: > This PR appears to work, but I'm not sure how to test the functionality extensively, beyond attempting login with 2FA and generating/using a magic link.

Here are some additional scenario you may want to consider (I tried to link pull requests with testing instructions that could be useful):

* Using Google to log in (aka. social login)
* Testing login in [other languages](https://github.com/Automattic/wp-calypso/pull/21118) (e.g. [French](https://wordpress.com/log-in/fr))
* Testing login with a redirect url
* Testing login from a [private site](https://github.com/Automattic/wp-calypso/pull/16590)
* Testing login from the desktop app
* Testing login with different [signup flows](https://github.com/Automattic/wp-calypso/pull/22983)

You could also test login flows with other OAuth clients such as [WooCommerce](https://github.com/Automattic/wp-calypso/pull/17699), or Jetpack.

> showing exactly the same login page for both logged in and logged out users.

I'm afraid I don't remember the use case for allowing a logged-in user to log into another account, maybe @scruffian will.

Comment (sgomes) [reply client/landing/login/store.js]: Thanks, @stephanethomas! Testing all of those seems like a really good idea.

@scruffian @westi Do you have any opinions on the use case for allowing a logged-in user to log into another account? We're trying to decide between @jsnajdr's two options, which I'm copying below:

> I see two good options:
> 
> 1. showing exactly the same login page for both logged in and logged out users. The masterbar doesn't indicate logged-in status. A downside is that already logged-in user is in a dead end here: there are no links to a logged-in view from the login page. The logged-in masterbar provided these links.
> 2. redirecting automatically to a logged-in home page. That's what Dropbox does, for example. The downside is that I can't easily login to another account. I need to go through the `/me` section and click `Sign Out` there. Few more clicks and you need to know where the "Sign Out" is.
> 
> Good amendment to option 1 would be detecting that I'm already logged in and offering to proceed or to use another account. That's what Google does. We don't need to bundle the huge `layout/logged-in` to achieve that: just issuing the `/me` request and storing the result in `state.currentUser` should be sufficient. 1-2KB is a good code size estimate.

**Comment by sgomes** [comment]
Still waiting for feedback on how to handle already logged-in users. Is there an approach we want to attempt in the meantime?

**Comment by sgomes** [comment]
Since we're still waiting for feedback, and in the interest of moving this PR forward, I implemented what I think is a reasonable solution to the already-logged-in scenario, by adding a prompt to the login form:

image

This should provide some indication to the user that they're already logged in, and filling in the form is unnecessary if they want to proceed with the same user, while also giving them the chance to log in as someone else.

Does this address your concerns, @jsnajdr ?

**Comment by sgomes** [comment]
The E2E test failures seem to be related to the "username or email" field sometimes losing focus, leaving the E2E tests unable to proceed. No matter how many times I try, I haven't been able to reproduce this situation locally at all.

@bsessions85 or anyone else familiar with them: is there something special about the failing E2E tests? Is there some special flag or config that modifies the login screen in a subtle way?

**Comment by bsessions85** [comment]
@sgomes I think what might be happening here is that we're trying to re-use login cookies in the tests and in the Plans tests it is conflicting with this new already logged in page. The login flow then expects it to be logged in already. I can't repro locally either, but I think the best option would be to set `useFreshLogin` to true for the plans tests.

**Comment by sgomes** [comment]
> @sgomes I think what might be happening here is that we're trying to re-use login cookies in the tests and in the Plans tests it is conflicting with this new already logged in page. The login flow then expects it to be logged in already. I can't repro locally either, but I think the best option would be to set `useFreshLogin` to true for the plans tests.

Thank you, @bsessions85 ! I'll see if I can do that.

**Comment by sgomes** [comment]
Rebased on master to pick up latest changes to `login-form.jsx`.

File: client/blocks/login/login-form.jsx
Comment (jsnajdr) [inline client/blocks/login/login-form.jsx]: The notice is not particularly beautiful and the text is very long. In #34541, with @davemart-in we added a subtle alternative action "or visit [some site]" under login header. It could be adapted like this:

Screenshot 2019-08-02 at 16 19 02

Adding a little gravatar next to my name would be a nice finishing touch. What do you think?

File: client/blocks/login/login-form.jsx
Comment (jsnajdr) [inline client/blocks/login/login-form.jsx]: The login URL can have a `redirect_to` parameter which specifies a different target URL than `/`. It should be used to continue as current user. The `login` Redux state contains this URL, and it's been already validated agains "open redirect" attacks.

File: client/controller/web-util.js
Comment (jsnajdr) [inline client/controller/web-util.js]: Creating this module could have been a zero-risk separate PR.

File: client/state/ui/reducer.js
Comment (jsnajdr) [inline client/state/ui/reducer.js]: Another opportunity for zero-risk separate PR. Makes the main diff much smaller.

File: client/blocks/login/login-form.jsx
Comment (sgomes) [reply client/blocks/login/login-form.jsx]: See, this is why I'm an engineer and not a designer 😄 I'll work on those changes, thanks!

File: client/blocks/login/login-form.jsx
Comment (sgomes) [reply client/blocks/login/login-form.jsx]: Oh, thanks, I'd forgotten about the URL parameter! And thanks for letting me know where to grab it from safely 👍

File: client/controller/web-util.js
Comment (sgomes) [reply client/controller/web-util.js]: You're right. I'll pull this out.

File: client/state/ui/reducer.js
Comment (sgomes) [reply client/state/ui/reducer.js]: I'll pull this out too.

File: client/controller/web-util.js
Comment (sgomes) [reply client/controller/web-util.js]: Done. See #35091

File: client/state/ui/reducer.js
Comment (sgomes) [reply client/state/ui/reducer.js]: Done! See #35092.

File: client/blocks/login/login-form.jsx
Comment (sgomes) [reply client/blocks/login/login-form.jsx]: The fields in `state.login` don't appear to be populated when first accessing the login page. As far as I can tell, they're only populated as the result of a `LOGIN_REQUEST`, which doesn't get issued until a login is attempted. The success response to this is what returns the sanitised URL.

I'm not sure what the best way of retrieving the sanitised redirect URL is, in this case. Any thoughts?

File: client/blocks/login/login-form.jsx
Comment (sgomes) [reply client/blocks/login/login-form.jsx]: Done!

image

**Comment by sgomes** [comment]
@jsnajdr : Should I extract the "continue as user" portion of this PR into a separate one?

File: client/blocks/continue-as-user/index.jsx
Comment (a8ci18n) [inline client/blocks/continue-as-user/index.jsx]: :information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

File: client/blocks/continue-as-user/index.jsx
Comment (jsnajdr) [inline client/blocks/continue-as-user/index.jsx]: I've been told by @lsl that `comment` is the param you want to use here. `context` is meant to be used to distinguish different meanings of the same string. E.g., "open" as verb vs. "open" as adjective.

Changing this will make the i18n bot happy, too 🙂

File: client/blocks/continue-as-user/index.jsx
Comment (jsnajdr) [inline client/blocks/continue-as-user/index.jsx]: Would `continue as {{userName/}}` better convey the meaning to translators?

File: client/blocks/continue-as-user/index.jsx
Comment (jsnajdr) [inline client/blocks/continue-as-user/index.jsx]: Adding a little gravatar would be nice 🙂

File: client/blocks/continue-as-user/index.jsx
Comment (jsnajdr) [inline client/blocks/continue-as-user/index.jsx]: This components should live in `client/blocks/login/continue-as-user.jsx`. It's not used anywhere else, unlike `VisitSite`, which is used also by `SiteSelector`.

**Comment by jsnajdr** [comment]
> Should I extract the "continue as user" portion of this PR into a separate one?

I think it can be landed separately, but let's also change the masterbar to always-logged-out in the same PR. Then the UI story makes sense: the same UI in both logged-in and logged-out state, and thanks to `ContinueAsUser`, you can still do everything you need to on a login page.

File: client/blocks/continue-as-user/index.jsx
Comment (sgomes) [reply client/blocks/continue-as-user/index.jsx]: You're right! Good bot 🤖

File: client/blocks/continue-as-user/index.jsx
Comment (sgomes) [reply client/blocks/continue-as-user/index.jsx]: Done 🙂 Thanks!

File: client/blocks/continue-as-user/index.jsx
Comment (sgomes) [reply client/blocks/continue-as-user/index.jsx]: Yes, very likely! Fixed.

File: client/blocks/continue-as-user/index.jsx
Comment (sgomes) [reply client/blocks/continue-as-user/index.jsx]: Added! Much nicer indeed 🙂 

image

File: client/blocks/continue-as-user/index.jsx
Comment (sgomes) [reply client/blocks/continue-as-user/index.jsx]: Done!

**Comment by sgomes** [comment]
> > Should I extract the "continue as user" portion of this PR into a separate one?
> 
> I think it can be landed separately, but let's also change the masterbar to always-logged-out in the same PR. Then the UI story makes sense: the same UI in both logged-in and logged-out state, and thanks to `ContinueAsUser`, you can still do everything you need to on a login page.

I made all the changes you suggested in this PR, to preserve some sort of sensible history, but I'll pull them out into a separate PR now. Once that's merged, I'll rebase this one.

**Comment by sgomes** [comment]
This PR is blocked on #35091, #35092, and #35309, which have all been extracted from code that originally started here.

**Comment by sgomes** [comment]
Squashed and rebased to pick up the changes in the already-rebased #35091 and #35092. Will do the same for #35309 once that's in.

**Comment by sgomes** [comment]
Rebased after "Continue as user" and style migration changes. This PR is a lot leaner and focused now! 🎉

File: client/layout/logged-out.jsx
Comment (sgomes) [inline client/layout/logged-out.jsx]: I think this is the best place for this import, but I'd love to hear your opinion, @jsnajdr!

Comment (jsnajdr) [reply client/layout/logged-out.jsx]: Not bad :+1: 🙂

Comment (sgomes) [reply client/layout/logged-out.jsx]: Thank you! 🙂

**Comment by jsnajdr** [comment]
Pushed a little fix that makes `a:visited` links in the login footer styled exactly the same as `a` (unvisited). With the new entrypoint, the order of styles is not deterministic anymore and it's path-dependent on previous navigations and interactions:

Either:
Screenshot 2019-09-27 at 13 46 21

Or:
Screenshot 2019-09-27 at 13 46 46

Notice how both `.wp-login__links a` and `a:visited` rules specify `color`, but a different color wins each time. That's fixed by efd37c0 by adding an explicit `wp-login__links a:visited` rule. That same commit also removes some redundant `text-decoration` and `body.is-section-signup` styles. These are already defined at other places and don't need to be specified again.

**Comment by jsnajdr** [comment]
One more remaining issue before this PR can be merged: #34770 by @andrewserong added some signup code (`getSignupProgress` and `setResumeAfterLogin`) to the `Login` block. That code reads and writes the Signup Redux state which is not available in the Login entrypoint. We'll need to refactor that to make Login 100% standalone.

**Comment by sgomes** [comment]
Rebased on master to pick up #36426

**Comment by jsnajdr** [comment]
There are a few more bugs, summarized in this screenshot:

Screenshot 2019-10-01 at 15 28 01

First, because the login handler no longer calls `renderLayout` from `client/boot/common`, the initial `ReactDom.hydrate` call renders over the placeholder HTML received from server. That doesn't end well, breaking the markup.

Second, the `EnvironmentBadge` doesn't render the `lib/abtest/test-helper` and `lib/preferences-helper` components. Also done in `setupMiddlewares` in `client/boot/common`.

I'll be looking into the first issue now.

**Comment by jsnajdr** [comment]
Pushed three commits now:
- reapply the CSS fixes from https://github.com/Automattic/wp-calypso/pull/33642#issuecomment-536461246 that got lost during @sgomes' rebase
- fix the broken `/log-in/es` SSR by always calling `ReactDom.render` instead of `ReactDom.hydrate`. `hydrate` requires the server markup to be identical to the initial client-side render, and our SSR is not in a good enough shape right now for that to work.
- added rendering of the abtests and preferences helpers. I had to add the `preferences` reducer to the Login store. Not very happy about that, but I don't see any other way to make the dev helpers work.

**Review by jsnajdr — APPROVED** [review]
I believe this is finally ready for merge :tada: Thanks @sgomes for the hard and long work!

**Comment by sgomes** [comment]
> * reapply the CSS fixes from [#33642 (comment)](https://github.com/Automattic/wp-calypso/pull/33642#issuecomment-536461246) that got lost during @sgomes' rebase

Sorry, I must have forgotten to pull your changes into my local branch 😞

**Comment by sgomes** [comment]
> I believe this is finally ready for merge 🎉 Thanks @sgomes for the hard and long work!

And thank you so much for all your invaluable help, @jsnajdr! Remember that I started off with your code, and we ended up with your fixes, so this is as much my work as yours 🙂

**Comment by Stojdza** [comment]
@sgomes I tried locally and every test which uses login form has failed. I'm not yet sure why, but I found that [this](https://github.com/Automattic/wp-calypso/pull/33642/commits/3422b31e58c766199cffb3ca07365e6fc390fd15#diff-eb93f2b0a7361af035359e5f30e5362aL31) line makes trouble. I know that we added it to improve e2e tests stability for logins. 

I've disabled it and let's see how tests will run now. Locally they went well 🤞

**Comment by kraftbj** [comment]
@sgomes I believe this changed has caused the redirect functionality fail, see #36518.

Example use case: Visit https://jetpack.com/contact-support/ logged out. Click the link to login.
Previously: It would redirect after login back to the page.
Now: Redirects to the Reader.

**Comment by sgomes** [comment]
> @sgomes I believe this changed has caused the redirect functionality fail, see #36518.
> 
> Example use case: Visit https://jetpack.com/contact-support/ logged out. Click the link to login.
> Previously: It would redirect after login back to the page.
> Now: Redirects to the Reader.

Thanks, @kraftbj !

To anyone following the thread, #36540 should fix the issue.

**Pull request merged**
